overlord/snapstate/handlers.go

```diff
@@ -3538,14 +3665,58 @@ func (m *SnapManager) doSetupAliases(t *state.Task, _ *tomb.Tomb) error {
 	}
 	snapName := snapsup.InstanceName()
 	curAliases := snapst.Aliases
+	autoDisabled := snapst.AutoAliasesDisabled
 
-	_, _, err = applyAliasesChange(snapName, autoDis, nil, snapst.AutoAliasesDisabled, curAliases, m.backend, doApply)
+	prune, oldAliases, oldAutoDisabled, err := shouldPruneOldAliases(t, snapst)
 	if err != nil {
 		return err
 	}
 
+	t.Set("old-aliases-pruned", prune)
+	// AliasesPending must be set to false before applying changes so that the
+	// undo for set-auto-aliases can revert aliases on disk if this task fails
+	// due to applyAliasesChange failing mid-way.
 	snapst.AliasesPending = false
```

Comment:
it is strange to set this in case of errors, as the disk and the state do not match then, we need to see if there's a minor tweak we can do that avoid needing this

only TestDoUndoSetupAliasesAutoErrorMidwayPruneOldAliases depends on this

it would be good to find a solution even it needs to tweak slightly undoRefreshAliases that doesn't involve marking AliasesPending false when it's not the case yet

Reply:
fixed